aten/src/ATen/core/PythonFallbackKernel.cpp

```diff
@@ -43,10 +43,23 @@ struct StashTLSOnEntryGuard {
   c10::impl::LocalDispatchKeySet saved_;
 };
 
-void pythonFallback(const c10::OperatorHandle& op, torch::jit::Stack* stack) {
+void pythonCompositeFallback(const c10::OperatorHandle& op, c10::DispatchKeySet ks, torch::jit::Stack* stack) {
+  c10::impl::ExcludeDispatchKeyGuard guard_(
+      c10::DispatchKeySet({c10::DispatchKey::Python, c10::DispatchKey::PythonComposite}));
+  op.callBoxed(stack);
+}
+
+void pythonFallback(const c10::OperatorHandle& op, c10::DispatchKeySet ks, torch::jit::Stack* stack) {
   TORCH_INTERNAL_ASSERT(tls_on_entry.has_value());
   // c10::impl::ForceDispatchKeyGuard dispatcher_guard(tls_on_entry.value());
   // StashTLSOnEntryGuard stash_guard;
+
+  // TODO: use guard semantics instead, see https://github.com/pytorch/pytorch/pull/74728
```

Comment:
How is the current PR working given that this is not restored after this call?

Reply:
exchange modifies it back to false. This makes it match @peterbell10's current PR although I'm presently trying to convince him to switch it to dynamic scoping.

Reply:
I am not sure this is sound though as the user has complete control of what happens between when you set that bool and when you unset it. A small modification of your code sample (disable inference mode inside) leads to pretty weird results:
```python
        called = []
        level = 0

        class SubTensor(torch.Tensor):
            __torch_function__ = torch._C._disabled_torch_function_impl

            @classmethod
            def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
                nonlocal level
                print(f"In {func} at {level} with {torch.is_inference_mode_enabled()}")
                called.append((func, level))
                level += 1
                try:
                    # I've seen what I wanted, enable decomps again
                    with torch.inference_mode(False):
                        return super().__torch_dispatch__(func, types, args, kwargs)
                finally:
                    level -= 1

        x = torch.randn(2, 3)
        y = torch.randn(5, 3)
        with torch.inference_mode(True):
            r = torch.nn.functional.linear(SubTensor(x), SubTensor(y))
```
Will generate
```
[(<OpOverload(op='aten.linear', overload='default')>, 0), (<OpOverload(op='aten.mm', overload='default')>, 1)]
```
Basically:
- The second linear is decomposed at the autograd level
- The torch_dispatch to transpose is the one that get skipped by the skip-one boolean
- matmul is decomposed at the autograd level
- mm is captured